### PR TITLE
Thisyahlen/FEQ-2045/chore: keep alive connection

### PR DESCRIPTION
**Pull request by thisyahlen-deriv**

### CODE REVIEW

File: src/context/api-context.tsx [hunk 1]
```diff
@@ -38,6 +38,8 @@ export const APIProvider = ({ children }: PropsWithChildren) => {
     const send: TSendFunction = (name, payload) => derivAPI.getActiveSocket().send({ [name]: 1, ...payload });
 
     useEffect(() => {
+        derivAPI.keepAlive();
```
Comment (balakrishna-deriv) [inline src/context/api-context.tsx:41]: IMO, we should not call the keepAlive inside a hook

File: src/context/api-context.tsx [hunk 2]
```diff
@@ -7,6 +7,8 @@ import { DerivAPILegacy } from '../client-library/deriv-api-client-legacy';
 
 const derivAPI = new DerivAPILegacy();
 
+derivAPI.keepAlive();
```
Comment (balakrishna-deriv) [inline src/context/api-context.tsx:10]: Ideally, we should be calling it after the wesocket is created

Comment (balakrishna-deriv) [reply src/context/api-context.tsx:10]: in deriv-api-legacy file

File: package.json [hunk 3]
```diff
@@ -7,6 +7,7 @@
         "dev": "vite",
         "build": "tsc && vite build",
         "prepare": "rm -rf ./dist && npm run build",
+        "watch": "vite build --watch",
```
[the remark this answers is not in the document]
Comment (thisyahlen-deriv) [reply package.json:10]: 😿

File: src/client-library/deriv-api-client-legacy.ts [hunk 4]
```diff
@@ -7,16 +7,46 @@ type WebsocketMap = Map<string, typeof DerivAPI>;
...
+
+    handleReconnect() {
+        const activeSocket = this.getActiveSocket();
+
+        if (!activeSocket || activeSocket.connection.readyState === WebSocket.CLOSED) {
+            const websocketURL = URLUtils.getWebsocketURL();
+            const webSocketInstance = new WebSocket(websocketURL);
+            const derivAPI = new DerivAPI({ connection: webSocketInstance });
+            this.websocketMap.set(websocketURL, derivAPI);
+            this.activeSocket = websocketURL;
```
Comment (balakrishna-deriv) [inline src/client-library/deriv-api-client-legacy.ts:34]: we need one more keepAlive call here

File: src/client-library/deriv-api-client-legacy.ts [hunk 5]
```diff
@@ -7,16 +7,48 @@ type WebsocketMap = Map<string, typeof DerivAPI>;
 export class DerivAPILegacy {
     activeSocket: string;
     websocketMap: WebsocketMap = new Map();
+    keepAliveIntervalId: NodeJS.Timeout | null = null;
```
Comment (jim-deriv) [inline src/client-library/deriv-api-client-legacy.ts:10]: Nit @thisyahlen-deriv, do we need this here if we're explicitly setting it to null? 
```suggestion
    keepAliveIntervalId: NodeJS.Timeout = null;
```

**Review by jim-deriv — APPROVED** [review]

File: src/client-library/deriv-api-client-legacy.ts [hunk 6]
```diff
@@ -7,16 +7,48 @@ type WebsocketMap = Map<string, typeof DerivAPI>;
...
+
+    handleReconnect() {
+        const activeSocket = this.getActiveSocket();
+
+        if (!activeSocket || activeSocket.connection.readyState === WebSocket.CLOSED) {
+            const websocketURL = URLUtils.getWebsocketURL();
+            const webSocketInstance = new WebSocket(websocketURL);
+            const derivAPI = new DerivAPI({ connection: webSocketInstance });
+            this.websocketMap.set(websocketURL, derivAPI);
+            this.activeSocket = websocketURL;
```
Comment (jim-deriv) [inline src/client-library/deriv-api-client-legacy.ts:34]: Also, this stuff seems to be repeated in the constructor, is it possible to create a helper method for them?
```suggestion
        if (!activeSocket || activeSocket.connection.readyState === WebSocket.CLOSED) {
            const websocketURL = URLUtils.getWebsocketURL();
            const webSocketInstance = new WebSocket(websocketURL);
            const derivAPI = new DerivAPI({ connection: webSocketInstance });
            this.websocketMap.set(websocketURL, derivAPI);
            this.activeSocket = websocketURL;
```

File: src/client-library/deriv-api-client-legacy.ts (hunk 5 above)
Comment (thisyahlen-deriv) [reply src/client-library/deriv-api-client-legacy.ts:10]: Yes @jim-deriv. I believe is not just setting keepAliveIntervalId to null, it's also declaring its type

File: src/client-library/deriv-api-client-legacy.ts (hunk 6 above)
Comment (thisyahlen-deriv) [reply src/client-library/deriv-api-client-legacy.ts:34]: can do, will refactor

**Review by balakrishna-deriv — APPROVED** [review]

File: src/client-library/deriv-api-client-legacy.ts (hunk 5 above)
Comment (jim-deriv) [reply src/client-library/deriv-api-client-legacy.ts:10]: Yeah, I just checked it, it's not inferred.

**Pull request merged**